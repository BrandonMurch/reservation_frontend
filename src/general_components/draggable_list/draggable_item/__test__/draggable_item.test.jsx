// Dependencies
import React from 'react';
import PropTypes from 'prop-types';
import {
  render, fireEvent, screen, createEvent,
} from '@testing-library/react';
import { create } from 'react-test-renderer';

// Components
import DraggableItem from '../index';

const DisplayComponent = ({ item }) => <h2>{item.text}</h2>;
DisplayComponent.propTypes = {
  item: PropTypes.shape({
    text: PropTypes.string.isRequired,
  }).isRequired,
};

describe('<DraggableItem />', () => {
  const props = {
    item: { text: 'Hello World' },
    index: 0,
    DisplayComponent,
    onDrop: jest.fn(),
    displayDroppable: true,
    setHovered: jest.fn(),
  };

  beforeEach(() => {
    render(
      <DraggableItem {...props} />,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const tree = create(
      <DraggableItem {...props} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call setData on dragStart', () => {
    const item = screen.getByText(/hello world/i);
    const mockSetData = jest.fn();
    const dragStartEvent = createEvent.dragStart(item);
    Object.assign(dragStartEvent, { dataTransfer: { setData: mockSetData } });
    fireEvent(item, dragStartEvent);
    expect(mockSetData).toHaveBeenCalled();
  });

  it('should call setHovered on dragEnter', () => {
    const item = screen.getByText(/hello world/i);
    fireEvent.dragEnter(item);
    expect(props.setHovered).toHaveBeenCalled();
  });

  it('should call setHovered on dragEnd', () => {
    const item = screen.getByText(/hello world/i);
    const dragEndEvent = createEvent.dragEnd(item);
    Object.assign(dragEndEvent, { dataTransfer: { dropEffect: 'none' } });
    fireEvent(item, dragEndEvent);
    expect(props.setHovered).toHaveBeenCalledWith(-1);
  });

  it('should call handleDrop on drop', () => {
    const item = screen.getByText(/hello world/i);
    fireEvent.drop(item);
    expect(props.onDrop).toHaveBeenCalled();
  });
});
