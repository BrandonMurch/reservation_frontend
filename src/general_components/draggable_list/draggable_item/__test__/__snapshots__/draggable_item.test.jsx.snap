// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DraggableItem /> should match snapshot 1`] = `
Array [
  <li
    className="droppable"
    onDragOver={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "display": "block",
      }
    }
  />,
  <li
    className="row"
    data-position={0}
    draggable={true}
    id="table 0"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "position": "relative",
      }
    }
  >
    <h2>
      Hello World
    </h2>
  </li>,
]
`;
